ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java

```diff
@@ -2434,15 +2434,20 @@ public static boolean isEmptyPath(Configuration job, Path dirPath) throws IOExce
 
   public static List<TezTask> getTezTasks(List<Task<? extends Serializable>> tasks) {
     List<TezTask> tezTasks = new ArrayList<TezTask>();
-    Set<Task<? extends Serializable>> visited = new HashSet<Task<? extends Serializable>>();
     if (tasks != null) {
-      getTezTasks(tasks, tezTasks, visited);
+      Set<Task<? extends Serializable>> visited = new HashSet<Task<? extends Serializable>>();
+      while (!tasks.isEmpty()) {
+        tasks = getTezTasks(tasks, tezTasks, visited);
+      }
     }
     return tezTasks;
   }
 
-  private static void getTezTasks(List<Task<? extends Serializable>> tasks, List<TezTask> tezTasks,
-      Set<Task<? extends Serializable>> visited) {
+  private static List<Task<? extends Serializable>> getTezTasks(
```

Comment:
Since the cleanup is a simple change, I think its worth getting that done as part of this change.
private static void getTasks(List<Task<? extends Serializable>> tasks, List<TezTask> tezTasks,
      Set<Task<? extends Serializable>> visited, Class requiredType) {
...
...
if( requiredType.instanceOf (task) && !resultTasks.contains(task)) {
... .
...

}